File: tutorials/integration/02-json-export/detect.py
#!/usr/bin/env python3
"""Export all SAF graphs to JSON for external tool integration.

Demonstrates the graph export pipeline: export each available graph
type and compute basic metrics (node count, edge count, max fan-out).

Usage:
    python detect.py
"""

import subprocess
import json
from collections import Counter
from pathlib import Path

from saf import Project


def _count_graph(graph_name: str, data: dict) -> tuple[int, int, Counter]:
    """Extract node count, edge count, and out-degree map for any graph type.

    All graph types now use the unified PropertyGraph format:
    {"schema_version": "0.1.0", "graph_type": "...",
     "nodes": [{"id": ..., "labels": [...], "properties": {...}}, ...],
     "edges": [{"src": ..., "dst": ..., "edge_type": "...", "properties": {...}}, ...]}
    """
    out_degree: Counter = Counter()

    nodes = data.get("nodes", [])
    edges = data.get("edges", [])

    for edge in edges:
        out_degree[edge.get("src", "")] += 1

    return len(nodes), len(edges), out_degree


def main() -> None:
    tutorial_dir = Path(__file__).parent
    source = tutorial_dir / "program.c"
    llvm_ir = tutorial_dir / "program.ll"

    subprocess.run(
        ["clang-18", "-S", "-emit-llvm", "-O0", "-g",
         "-o", str(llvm_ir), str(source)],
        check=True,
    )

    proj = Project.open(str(llvm_ir))
    graphs = proj.graphs()

    available = graphs.available()
    print(f"Available graph types: {available}")

    for graph_name in available:
        print(f"\n{'='*50}")
        print(f"Graph: {graph_name}")
        print(f"{'='*50}")

        data = graphs.export(graph_name)

        n_nodes, n_edges, out_degree = _count_graph(graph_name, data)

        print(f"  Nodes: {n_nodes}")
        print(f"  Edges: {n_edges}")

        if out_degree:
            max_node, max_deg = out_degree.most_common(1)[0]
            short = str(max_node)[:20] + "..." if len(str(max_node)) > 20 else str(max_node)
            print(f"  Max fan-out: {max_deg} (node {short})")

        # Save to file
        out_file = tutorial_dir / f"{graph_name}.json"
        with open(out_file, "w") as f:
            json.dump(data, f, indent=2, default=str)
        print(f"  Saved to: {out_file.name}")


if __name__ == "__main__":
    main()

File: tutorials/integration/02-json-export/detect.rs
// Export all SAF graphs to JSON using the Rust API.
//
// Demonstrates building and exporting each graph type: CFG, CallGraph,
// DefUse, and ValueFlow.
//
// Build and run:
//   cargo run --features llvm-18 --example detect_json_export

use std::path::PathBuf;
use std::process::Command;

use saf_analysis::callgraph::CallGraph;
use saf_analysis::cfg::Cfg;
use saf_analysis::defuse::DefUseGraph;
use saf_analysis::{ValueFlowBuilder, ValueFlowConfig};
use saf_core::config::Config;
use saf_frontends::api::Frontend;
use saf_frontends::llvm::LlvmFrontend;

fn main() {
    let tutorial_dir = PathBuf::from("tutorials/integration/02-json-export-pipeline");
    let source = tutorial_dir.join("program.c");
    let llvm_ir = tutorial_dir.join("program.ll");

    let status = Command::new("clang-18")
        .args(["-S", "-emit-llvm", "-O0", "-g",
               "-o", llvm_ir.to_str().unwrap(), source.to_str().unwrap()])
        .status()
        .expect("failed to run clang-18");
    assert!(status.success());

    let frontend = LlvmFrontend::new();
    let config = Config::default();
    let bundle = frontend.ingest(&[llvm_ir.as_path()], &config)
        .expect("failed to load LLVM IR");
    let module = bundle.module;

    // Build all graphs
    let cfg = Cfg::build(&module);
    let callgraph = CallGraph::build(&module);
    let defuse = DefUseGraph::build(&module);

    let vf_config = ValueFlowConfig::default();
    let builder = ValueFlowBuilder::new(&vf_config, &module, &defuse, &callgraph, None);
    let vf = builder.build();

    // Export and report
    let cfg_export = cfg.export();
    println!("CFG: {} nodes, {} edges", cfg_export.nodes.len(), cfg_export.edges.len());

    let cg_export = callgraph.export();
    println!("CallGraph: {} nodes, {} edges", cg_export.nodes.len(), cg_export.edges.len());

    let du_export = defuse.export();
    println!("DefUse: {} nodes, {} edges", du_export.nodes.len(), du_export.edges.len());

    let vf_export = vf.export();
    println!("ValueFlow: {} nodes, {} edges", vf_export.nodes.len(), vf_export.edges.len());
}
